src/type-logic.ts

```diff
@@ -398,6 +402,29 @@ export function typeFromExpr(expr: Expr, biblio: Biblio): Type {
       }
       const calleeName = callee[0].contents;
 
+      // special case: `Completion` is identity on completions
+      if (expr.name === 'call' && calleeName === 'Completion' && expr.arguments.length === 1) {
+        const inner = typeFromExpr(expr.arguments[0], biblio);
+        if (!isCompletion(inner)) {
+          // probably unknown, we might as well refine to "some completion"
+          return {
+            kind: 'union',
+            of: [
+              { kind: 'normal completion', of: { kind: 'unknown' } },
+              { kind: 'abrupt completion' },
+            ],
+          };
+        }
+      }
+      // special case: `NormalCompletion` wraps its input
+      if (
+        expr.name === 'call' &&
+        calleeName === 'NormalCompletion' &&
+        expr.arguments.length === 1
```

Comment:
I think these arguments length checks should be assertions, not tests.

Reply:
Done. I don't think it was worth my time to make that change, but too late now.